src/NHibernate/Engine/ActionQueue.cs

```diff
@@ -505,106 +507,176 @@ private class InsertActionSorter
+				}
+
+				public override int GetHashCode()
+				{
+					return EntityName.GetHashCode();
+				}
+			}
+
+			// _orderedBatches and _entityNameBatchIdentifier replace _latestBatch from Hibernate implementation.
+			private readonly List<BatchIdentifier> _orderedBatches = new List<BatchIdentifier>();
+			private readonly Dictionary<string, BatchIdentifier> _entityNameBatchIdentifier = new Dictionary<string, BatchIdentifier>();
+			private readonly Dictionary<object, BatchIdentifier> _entityBatchIdentifier;
 
 			// the map of batch numbers to EntityInsertAction lists
-			private readonly Dictionary<int, List<EntityInsertAction>> _actionBatches = new Dictionary<int, List<EntityInsertAction>>();
+			private readonly Dictionary<BatchIdentifier, List<EntityInsertAction>> _actionBatches = new Dictionary<BatchIdentifier, List<EntityInsertAction>>();
 
 			public InsertActionSorter(ActionQueue actionQueue)
 			{
 				_actionQueue = actionQueue;
 
 				//optimize the hash size to eliminate a rehash.
-				_entityBatchNumber = new Dictionary<object, int>(actionQueue.insertions.Count + 1);
+				_entityBatchIdentifier = new Dictionary<object, BatchIdentifier>(actionQueue.insertions.Count + 1);
 			}
 
 			public void Sort()
 			{
 				// the list of entity names that indicate the batch number
 				foreach (EntityInsertAction action in _actionQueue.insertions)
 				{
-					// remove the current element from insertions. It will be added back later.
-					var entityName = action.EntityName;
-
 					// the entity associated with the current action.
 					var currentEntity = action.Instance;
 
-					var batchNumber = GetBatchNumber(action, entityName);
-					_entityBatchNumber[currentEntity] = batchNumber;
-					AddToBatch(batchNumber, action);
+					BatchIdentifier batchIdentifier;
+					if (!_entityNameBatchIdentifier.TryGetValue(action.EntityName, out batchIdentifier))
+					{
+						batchIdentifier = new BatchIdentifier(action.EntityName);
+						_entityNameBatchIdentifier[action.EntityName] = batchIdentifier;
+						_orderedBatches.Add(batchIdentifier);
+					}
+					AddParentChildEntityNames(action, batchIdentifier);
+					_entityBatchIdentifier.Add(currentEntity, batchIdentifier);
+					AddToBatch(batchIdentifier, action);
 				}
 				_actionQueue.insertions.Clear();
 
-				// now rebuild the insertions list. There is a batch for each entry in the name list.
-				for (var i = 0; i < _actionBatches.Count; i++)
+				for (var i = 0; i < _orderedBatches.Count; i++)
 				{
-					var batch = _actionBatches[i];
-					foreach (var action in batch)
+					var batchIdentifier = _orderedBatches[i];
+					var entityName = batchIdentifier.EntityName;
+
+					//Make sure that child entries are not before parents
+					for (var j = i - 1; j >= 0; j--)
 					{
-						_actionQueue.insertions.Add(action);
+						var prevBatchIdentifier = _orderedBatches[j];
+						if (prevBatchIdentifier.ParentEntityNames.Contains(entityName))
+						{
+							_orderedBatches.RemoveAt(i);
```

Comment:
Hum, using a `LinkedList` could bring us back to O(n²) and avoid this kind of issues, since we will be able to work with nodes instead of indexes. It should way more readable, and so easier to write it correctly.

I will give it a try.